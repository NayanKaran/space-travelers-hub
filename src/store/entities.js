import { combineReducers } from '@reduxjs/toolkit';
import rocketsReducer from './rockets';
import missionsReducer from './missions';

export default combineReducers({
  rockets: rocketsReducer,
  missions: missionsReducer,
});
