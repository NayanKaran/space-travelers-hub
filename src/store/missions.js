import axios from 'axios';

// Actions
const FETCH_DATA = 'FETCH_DATA_FROM_API';
const TOGGLE_RESERVATION = 'TOGGLE_RESERVATION';

const missions = [];

// reducer
export default function missionsReducer(state = missions, action) {
  switch (action.type) {
    case FETCH_DATA: {
      return action.payload;
    }

    case TOGGLE_RESERVATION: {
      return state.map((mission) => {
        if (mission.mission_id !== action.id) {
          return mission;
        }
        return { ...mission, reserved: !mission.reserved };
      });
    }

    default:
      return state;
  }
}

// Action Creators
const fetchMission = (payload) => ({
  type: FETCH_DATA,
  payload,
});

export const joinMission = (id) => ({
  type: TOGGLE_RESERVATION,
  id,
});

/* eslint-disable camelcase */

// mission api
const url = 'https://api.spacexdata.com/v3/missions';

export const fetchData = () => async (dispatch) => {
  const response = await axios.get(url);
  const data = Object.entries(response.data).map((item) => {
    const {
      mission_id, mission_name, description,
    } = item[1];
    return {
      mission_id,
      mission_name,
      description,
      reserved: false,
    };
  });
  dispatch(fetchMission(data));
};
