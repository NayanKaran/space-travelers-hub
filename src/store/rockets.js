import { createSlice } from '@reduxjs/toolkit';

const slice = createSlice({
  name: 'rockets',
  initialState: [],
  reducers: {
    checkStatus: () => 'Under construction',
  },
});

export const { checkStatus } = slice.actions;

export default slice.reducer;
