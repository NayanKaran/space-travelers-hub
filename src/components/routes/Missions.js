import React, { useEffect } from 'react';
import { useSelector, useDispatch } from 'react-redux/es/exports';
import { joinMission, fetchData } from '../../store/missions';
import styles from './Missions.module.css';

const Missions = () => {
  const state = useSelector((state) => state.entities.missions);
  const dispatch = useDispatch();
  useEffect(() => {
    dispatch(fetchData());
  }, [dispatch]);

  const handleMission = (id) => {
    dispatch(joinMission(id));
  };

  return (
    <main id="missions">
      <table className={styles.missionTable}>
        <thead>
          <tr>
            <th>Mission</th>
            <th>Description</th>
            <th>Status</th>
          </tr>
        </thead>
        <tbody>
          {state.map((item) => (
            /* eslint-disable camelcase */
            <tr key={item.mission_id} className={styles.tableRow}>
              <td>{item.mission_name}</td>
              <td>{item.description}</td>
              <td>
                <button type="button" className={item.reserved ? styles.activeMember : styles.notActive}>{item.reserved ? 'Active member' : 'NOT A MEMBER'}</button>
              </td>
              <td>
                <button type="button" className={item.reserved ? styles.leaveMissionBtn : styles.joinMissionBtn} onClick={() => handleMission(item.mission_id)}>
                  {item.reserved ? 'Leave ' : 'Join '}
                  Mission
                </button>
              </td>
            </tr>
          ))}
        </tbody>
      </table>
    </main>
  );
};

export default Missions;
