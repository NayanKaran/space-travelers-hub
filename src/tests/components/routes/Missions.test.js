import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import Missions from '../../../components/routes/Missions';
import store from '../../../store/configureStore';

describe('Testing Missions component', () => {
  it('Testing with snapshot', () => {
    const tree = renderer
      .create(
        <Provider store={store}>
          <Missions />
        </Provider>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
