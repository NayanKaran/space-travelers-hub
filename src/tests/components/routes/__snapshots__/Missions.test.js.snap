// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Missions component Testing with snapshot 1`] = `
<main
  id="missions"
>
  <table
    className="missionTable"
  >
    <thead>
      <tr>
        <th>
          Mission
        </th>
        <th>
          Description
        </th>
        <th>
          Status
        </th>
      </tr>
    </thead>
    <tbody />
  </table>
</main>
`;
